winml/test/concurrency/ConcurrencyTests.cpp

```diff
@@ -0,0 +1,279 @@
+    std::vector<EvaluationUnit> evaluation_units(num_units, EvaluationUnit());
+    auto ifv = FileHelpers::LoadImageFeatureValue(L"kitten_224.png");
+
+    // same model, different session
+    auto model = LearningModel::LoadFromFilePath(FileHelpers::GetModulePath() + L"model.onnx");
+    for (unsigned int i = 0; i < num_units; ++i)
+    {
+        evaluation_units[i].model = model;
+        evaluation_units[i].session = LearningModelSession(evaluation_units[i].model);
+        evaluation_units[i].binding = LearningModelBinding(evaluation_units[i].session);
+        evaluation_units[i].binding.Bind(L"data_0", ifv);
+    }
+
+    RunAsync(evaluation_units);
+    std::vector<uint32_t> indices(num_units, TABBY_CAT_INDEX);
+    VerifyEvaluation(evaluation_units, indices);
+}
+
+// Run evaluations with same session (and model), with different bindings
+void EvalAsyncDifferentBindings()
+{
+    unsigned int num_units = 2;
+    std::vector<EvaluationUnit> evaluation_units(num_units, EvaluationUnit());
+
+    std::vector<ImageFeatureValue> ifvs = {FileHelpers::LoadImageFeatureValue(L"kitten_224.png"),
+                                           FileHelpers::LoadImageFeatureValue(L"fish.png")};
+
+    // same session, different binding
+    auto model = LearningModel::LoadFromFilePath(FileHelpers::GetModulePath() + L"model.onnx");
+    auto session = LearningModelSession(model);
+    for (unsigned int i = 0; i < num_units; ++i)
+    {
+        evaluation_units[i].model = model;
+        evaluation_units[i].session = session;
+        evaluation_units[i].binding = LearningModelBinding(evaluation_units[i].session);
+        evaluation_units[i].binding.Bind(L"data_0", ifvs[i]);
+    }
+
+    RunAsync(evaluation_units);
+    VerifyEvaluation(evaluation_units, { TABBY_CAT_INDEX, TENCH_INDEX });
+}
+
+winrt::Windows::AI::MachineLearning::ILearningModelFeatureDescriptor UnusedCreateFeatureDescriptor(
+    std::shared_ptr<onnxruntime::Model> model,
+    const std::wstring& name,
+    const std::wstring& description,
+    bool is_required,
+    const ::onnx::TypeProto *type_proto);
+
+// Get random number in interval [1,max_number]
+unsigned int getRandomNumber(unsigned int max_number) {
+    return std::rand() % max_number + 1;
+}
+
+void MultiThreadLoadModel() {
+    // load same model
+    auto path = FileHelpers::GetModulePath() + L"model.onnx";
+    unsigned int num_threads = getRandomNumber(MAX_THREADS);
+    ThreadPool pool(num_threads);
+    try {
+        for (unsigned int i = 0; i < num_threads; ++i) {
+            pool.SubmitWork([&path]() {
+                auto model = LearningModel::LoadFromFilePath(path);
+                std::wstring name(model.Name());
+                WINML_EXPECT_EQUAL(name, L"squeezenet_old");
+            });
+        }
+    }
+    catch (...) {
+        WINML_LOG_ERROR("Failed to load model concurrently.");
+    }
+}
+
+// Create different sessions for each thread, and evaluate
+void MultiThreadSession()
```

Comment:
This is likely beyond the scope of this PR, but this test validates multithreadsession, but ends up testing multi-threaded e2e.

It is possible that 1 thread is in evaluate while another is in bind, and another is creating the session, and another is validating the results.

We don't actually forcefully validate the concurrent calls of these API combinations.


This can probably be accomplished by keeping the same method structure, but setting condition variables before each call that are listened to by the main test thread, and waiting on a cv that the main thread triggers to ensure concurrent execution.